homeassistant/components/tuya/__init__.py

```diff
@@ -0,0 +1,296 @@
+        _LOGGER.debug("tuya.__init__.exist_xor_cache-->True")
+        key_iv = aes.xor_decrypt(init_entry_data[XOR_KEY], init_entry_data[KEY_KEY])
+        cbc_key = key_iv[0:16]
+        cbc_iv = key_iv[16:32]
+        decrpyt_str = aes.cbc_decrypt(cbc_key, cbc_iv, init_entry_data[AES_ACCOUNT_KEY])
+        entry_data = aes.json_to_dict(decrpyt_str)
+    else:
+        # if not exist xor cache, use old account info
+        _LOGGER.debug("tuya.__init__.exist_xor_cache-->False")
+        entry_data = init_entry_data
+        cbc_key = aes.random_16()
+        cbc_iv = aes.random_16()
+        access_id = init_entry_data[CONF_ACCESS_ID]
+        access_id_entry = aes.cbc_encrypt(cbc_key, cbc_iv, access_id)
+        merged_data = cbc_key + cbc_iv
+        c_xor_entry = aes.xor_encrypt(merged_data, access_id_entry)
+        # account info encrypted with AES-CBC
+        user_input_encrpt = aes.cbc_encrypt(
+            cbc_key, cbc_iv, json.dumps(dict(init_entry_data))
+        )
+        # update old account info
+        hass.config_entries.async_update_entry(
+            entry,
+            data={
+                AES_ACCOUNT_KEY: user_input_encrpt,
+                XOR_KEY: c_xor_entry,
+                KEY_KEY: access_id_entry,
+            },
+        )
+    return entry_data
+
+
+async def _init_tuya_sdk(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    init_entry_data = entry.data
+    # decrypt or encrypt entry info
+    entry_data = entry_decrypt(hass, entry, init_entry_data)
+    project_type = ProjectType(entry_data[CONF_PROJECT_TYPE])
+    api = TuyaOpenAPI(
+        entry_data[CONF_ENDPOINT],
+        entry_data[CONF_ACCESS_ID],
+        entry_data[CONF_ACCESS_SECRET],
+        project_type,
+    )
+
+    api.set_dev_channel("hass")
+
+    if project_type == ProjectType.INDUSTY_SOLUTIONS:
+        response = await hass.async_add_executor_job(
+            api.login, entry_data[CONF_USERNAME], entry_data[CONF_PASSWORD]
+        )
+    else:
+        response = await hass.async_add_executor_job(
+            api.login,
+            entry_data[CONF_USERNAME],
+            entry_data[CONF_PASSWORD],
+            entry_data[CONF_COUNTRY_CODE],
+            entry_data[CONF_APP_TYPE],
+        )
+
+    if response.get("success", False) is False:
+        _LOGGER.error("Tuya login error response: %s", response)
+        return False
+
+    tuya_mq = TuyaOpenMQ(api)
+    tuya_mq.start()
+
+    device_manager = TuyaDeviceManager(api, tuya_mq)
+
+    # Get device list
+    home_manager = TuyaHomeManager(api, tuya_mq, device_manager)
+    await hass.async_add_executor_job(home_manager.update_device_cache)
+    hass.data[DOMAIN][TUYA_HOME_MANAGER] = home_manager
+
+    class DeviceListener(TuyaDeviceListener):
+        """Device Update Listener."""
+
+        def update_device(self, device: TuyaDevice):
+            for ha_device in hass.data[DOMAIN][TUYA_HA_DEVICES]:
+                if ha_device.tuya_device.id == device.id:
```

Comment:
This is a very expensive lookup with O(N) complexity. I suggest you convert `TUYA_HA_DEVICES` into a dictionary so you can have O(1) lookups. You can maintain a list per device since there can be multiple entities.

Also, these are `Entity` objects, not devices. Update the constant name. 

And it's not good for an integration to directly interact with entity objects. Because a created entity object might not actually be added to Home Assistant. Instead, let TuyaEntities register themselves when they are added to HA and let them remove themselves when they are removed.

Add this to the Tuya entity base class:

```python
    async def async_added_to_hass(self):
        await super().async_added_to_hass()
        entities = self.hass.data[DOMAIN][TUYA_HA_DEVICES].setdefault(self.tuya_device.id, [])
        entities.append(self)
        self.async_on_remove(lambda: entities.remove(self))
```

Reply:
We should probably use our dispatch helper to signal the entity from outside instead so we shouldn't need to store the entity in a shared container at all.